File: packages/core/src/sampling.ts
// Client-side error sampling. Mirrors Reporter#sampled_out?: a rate >= 1 keeps everything,
// <= 0 drops everything, otherwise keep with probability `rate`. The rng is injectable so
// tests are deterministic.
export function sampledOut(rate: number, rng: () => number = Math.random): boolean {
  if (rate >= 1) return false;
  if (rate <= 0) return true;
  return rng() > rate;
}

File: python/dispatchitapp/src/dispatchitapp/sampling.py
"""Client-side error sampling. Mirrors Reporter#sampled_out?."""

from __future__ import annotations

import random
from typing import Callable


def sampled_out(rate: float, rng: Callable[[], float] = random.random) -> bool:
    if rate >= 1:
        return False
    if rate <= 0:
        return True
    return rng() > rate
